architecture/2024-05-radius-on-dapr.md

```diff
@@ -0,0 +1,324 @@
+
+This is our existing precedent for a dependency that's installed with Radius. We should not deviate from this or add additional options without getting user feedback.
+
+We'll need to configure the Dapr [annotations](https://docs.dapr.io/operations/configuration/configuration-overview/#sidecar-configuration) on UCP, ApplicationsRP, and the DE to enable Dapr.
+
+We should also use our existing metrics and otel configuration settings to configure [Dapr's telemetry features](https://docs.dapr.io/operations/configuration/configuration-overview/#tracing).
+
+We should not add a dependency from our Helm chart onto Dapr's Helm chart. Dependencies in Helm aren't widely used because they are inflexible. It's better if users can manage Dapr on their own terms.
+
+#### Default components
+
+We need to choose a default state store and pub/sub to be included with `rad install` and `rad init`. 
+
+On paper, best option sounds like Redis, because it's both a state store and pub/sub. This is also ways `dapr init` installs by default. However, we're likely to leverage transactions heavily in the implementation of the database. Redis only has [partial support](https://docs.dapr.io/reference/components-reference/supported-state-stores/setup-redis/) for transactions, so I propose that we avoid it.
+
+Our criteria for deciding the best components should look something like:
+
+- Supports the features we need.
+- Open-source and self-hostable.
+- High level of confidence in the quality of documentation and support.
+- Low runtime overhead and simple management.
+
+**State store:** From this [list](https://docs.dapr.io/reference/components-reference/supported-state-stores/) my shortlist is CockroachDB, etcd, MongoDB, MySql/MariaDB, and Postgres.
+
+**Pub/sub:** From this [list](https://docs.dapr.io/reference/components-reference/supported-pubsub/) my shortlist is Kafka, JetStream, RabbitMQ, and Redis.
+
+I have a slight preference towards Postgres and RabbitMQ due to my familiarity, but we should do more investigation before deciding.
+
+---
+
+Whichever options we choose, we'll add them to our Helm chart with a new option `global.dapr.defaultComponents=true` as an opt-in. Our `rad init` and `rad install` commands will set this option. 
+
+I feel better about opt-in vs out-out for the Helm Chart. It's very typical that Helm Charts require additional settings or configuration when dependencies are involved. 
+
+These components should be configured to balance resource usage and complexity versus production readiness. We want to do something reasonable, but we also don't expect users to use our default components for production.
+
+#### State store
+
+Our task is to implement the [`StorageClient` API](https://github.com/radius-project/radius/blob/main/pkg/ucp/store/client.go#L25) on top of the Dapr state store API. Since we'll be keeping the existing store API this change should have no impact on the rest of the codebase.
```

Comment:
We can easily switch between CosmosDB and etcd. We can go with something like that.

Reply:
👍